File: solutions/prospective/algorithms/patterns/py/hashmap_frequency_template.py
"""
HashMap Frequency Template

Standard counting pattern using dict or Counter.
Often the first step in anagrams, most frequent elements, etc.
"""
=== MEMO START ===
def count_frequencies(nums):
    counts = {}
    for num in nums:
        counts[num] = counts.get(num, 0) + 1
    return counts

# Or manually:
# counts = {}
# for x in s:
#     if x not in counts:
#         counts[x] = 0
#     counts[x] += 1

File: solutions/prospective/algorithms/patterns/js/hashmap_frequency_template.js
/**
 * HashMap Frequency Template
 * 
 * Standard counting pattern using object or Map.
 * Often the first step in anagrams, most frequent elements, etc.
 */
// === MEMO START ===
function countFrequencies(nums) {
    const counts = {};
    for (const num of nums) {
        counts[num] = (counts[num] || 0) + 1;
    }
    return counts;
}

// Or using Map:
// function countFrequencies(nums) {
//     const counts = new Map();
//     for (const num of nums) {
//         counts.set(num, (counts.get(num) || 0) + 1);
//     }
//     return counts;
// }

